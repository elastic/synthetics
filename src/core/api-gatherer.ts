/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { request as apiRequest } from 'playwright-core';
import { log } from './logger';
import { APIDriver, RunOptions } from '../common_types';
import { APIPluginManager } from '../plugins/api-plugin-manager';

/**
 * Purpose of the Gatherer is to set up the necessary browser driver
 * related capabilities for the runner to run all journeys
 */
export class APIGatherer {
  static pluginManager: APIPluginManager;

  static async setupDriver(options: RunOptions) {
    log('Gatherer: setup browser');
    const { playwrightOptions } = options;

    // Register sig int handler to close the browser
    process.on('SIGINT', async () => {
      await APIGatherer.stop();
      process.exit(130);
    });

    const request = await apiRequest.newContext({ ...playwrightOptions });
    return { request };
  }

  /**
   * Starts recording all events related to the v8 devtools protocol
   * https://chromedevtools.github.io/devtools-protocol/v8/
   */

  static async beginRecording(driver: APIDriver, options: RunOptions) {
    log('Gatherer: started recording');
    const { network, metrics } = options;
    APIGatherer.pluginManager = new APIPluginManager(driver);
    APIGatherer.pluginManager.registerAll(options);
    const plugins = [await APIGatherer.pluginManager.start('browserconsole')];
    network && plugins.push(await APIGatherer.pluginManager.start('network'));
    metrics &&
      plugins.push(await APIGatherer.pluginManager.start('performance'));
    await Promise.all(plugins);
    return APIGatherer.pluginManager;
  }

  static async endRecording() {
    log('Gatherer: ended recording');
    await APIGatherer.pluginManager.unregisterAll();
  }

  static async dispose(driver: APIDriver) {
    log(`Gatherer: closing all contexts`);
    await driver.request.dispose();
  }

  static async stop() {
    log(`Gatherer: closing browser`);
    // if (Gatherer.request) {
    //   await Gatherer.browser.close();
    //   Gatherer.browser = null;
    // }
  }
}
