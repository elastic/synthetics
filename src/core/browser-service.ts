/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { URL } from 'url';
import { createServer } from 'http';
import { createProxyServer } from 'http-proxy';
import { chromium } from 'playwright-chromium';
import { log } from './logger';

export class BrowserService {
  closeCallbacks = [];
  proxyServer = createServer();

  constructor(private options = { port: 9322 }) {}

  init() {
    const proxy = createProxyServer();

    this.proxyServer.on('upgrade', async (req, socket, head) => {
      const browserServer = await chromium.launchServer({ headless: true });
      const wsEndpoint = browserServer.wsEndpoint();
      const { pathname, origin } = new URL(wsEndpoint);
      req.url = pathname;
      log(`New browser: ${wsEndpoint}`);
      proxy.ws(req, socket, head, { target: origin });
      const closeBrowser = async () => {
        const index = this.closeCallbacks.indexOf(closeBrowser);
        if (index >= 0) {
          this.closeCallbacks.splice(index, 1);
        }
        await browserServer.close();
        log(`Socket closed: ${wsEndpoint}`);
      };
      this.closeCallbacks.push(closeBrowser);
      socket.on('close', closeBrowser);
      socket.on('error', closeBrowser);
    });

    this.proxyServer.listen(this.options.port, () => {
      log(`Listening on port: ${this.options.port}`);
    });
  }

  async dispose() {
    for (let i = 0; i < this.closeCallbacks.length; i++) {
      const callback = this.closeCallbacks[i];
      await callback();
    }
    this.closeCallbacks = [];

    await new Promise(resolve => {
      this.proxyServer.close(resolve);
    });
  }
}
