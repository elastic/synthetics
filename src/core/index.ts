/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { Journey, JourneyCallback, JourneyOptions } from '../dsl';
import Runner from './runner';
import { VoidCallback } from '../common_types';
import { log } from './logger';

export const runner = new Runner();

export const journey = (
  options: JourneyOptions | string,
  callback: JourneyCallback
) => {
  log(`register journey: ${options}`)
  if (typeof options === 'string') {
    options = { name: options, id: options };
  }
  const j = new Journey(options, callback);
  runner.addJourney(j);
  return j;
};

export const step = (name: string, callback: VoidCallback) => {
  log(`register step: ${name}`)
  return runner.currentJourney?.addStep(name, callback);
};

export const beforeAll = (callback: VoidCallback) => {
  runner.addHook('beforeAll', callback);
};

export const afterAll = (callback: VoidCallback) => {
  runner.addHook('afterAll', callback);
};

export const before = (callback: VoidCallback) => {
  if (!runner.currentJourney) {
    throw new Error('before is called outside of the journey context');
  }
  return runner.currentJourney.addHook('before', callback);
};

export const after = (callback: VoidCallback) => {
  if (!runner.currentJourney) {
    throw new Error('after is called outside of the journey context');
  }
  return runner.currentJourney.addHook('after', callback);
};
