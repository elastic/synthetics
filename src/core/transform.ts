/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { addHook } from 'pirates';
import {
  transformSync,
  Loader,
  CommonOptions,
  TransformOptions,
} from 'esbuild';
import path from 'path';
import sourceMapSupport from 'source-map-support';

// Cache that holds the sourcemap content for each file
const sourceMaps: Map<string, string> = new Map();

// Register the source-map-support library to resolve the sourcemaps
// for the files that are transpiled by esbuild
/**
 * Register the source-map-support library to resolve the sourcemaps
 * for the files that are transpiled by esbuild and cache the sourcemaps
 * for each file.
 *
 * Caching the sourcemap file is important, as it is used by the Synthetics agent
 * to resolve the stack traces of journey, monitor and step functions.
 */
sourceMapSupport.install({
  environment: 'node',
  handleUncaughtExceptions: false,
  retrieveSourceMap(source) {
    if (!sourceMaps.has(source)) return null;
    return {
      map: JSON.parse(sourceMaps.get(source)),
      url: source,
    };
  },
});

/**
 * Default list of files and corresponding loaders we support
 * while pushing project based monitors
 */
const LOADERS: Record<string, Loader> = {
  '.ts': 'ts',
  '.js': 'js',
  '.mjs': 'js',
  '.cjs': 'js',
};

const getLoader = (filename: string) => {
  const ext = path.extname(filename);
  return LOADERS[ext] || 'default';
};

export function commonOptions(): CommonOptions {
  /**
   * We are not minifying the code as we want it to be in sync with previous
   * transformation phase and introduce it later.
   */
  return {
    minify: false,
    minifyIdentifiers: false,
    minifySyntax: false,
    minifyWhitespace: false,
    sourcemap: 'both',
    sourcesContent: false,
    platform: 'node',
    logLevel: 'silent',
    format: 'cjs',
    target: `node${process.version.slice(1)}`,
  };
}

/**
 * Transform the given code using esbuild and save the corresponding
 * map file in memory to be retrived later.
 */
export function transform(
  code: string,
  filename: string,
  options: TransformOptions = {}
) {
  const result = transformSync(code, {
    ...commonOptions(),
    sourcefile: filename,
    loader: getLoader(filename),
    /**
     * Add this only for the transformation phase, using it on
     * bundling phase would disable tree shaking and uncessary bloat
     *
     * Ensures backwards compatability with tsc's implicit strict behaviour
     */
    tsconfigRaw: {
      compilerOptions: {
        alwaysStrict: true,
      },
    },
    ...options,
  });

  const warnings = result.warnings;
  if (warnings && warnings.length > 0) {
    for (const warning of warnings) {
      console.log(warning.location);
      console.log(warning.text);
    }
  }

  /**
   * Cache the sourcemap contents in memory, so we can look it up
   * later when we try to resolve the sourcemap for a given file.
   *
   * Every time the synthetics agent is started, we register the source-map-support
   * library to resolve the sourcemaps for the files that are transpiled by esbuild.
   */
  if (result.map) {
    sourceMaps.set(filename, result.map);
  }
  return result;
}

/**
 * Install the pirates hook to transform the code on the fly
 * for all of the imported files.
 */
export function installTransform() {
  const revertPirates = addHook(
    (source: string, filename: string) => {
      const { code } = transform(source, filename);
      return code;
    },
    { exts: ['.ts', '.js', '.mjs', '.cjs'] }
  );

  return () => {
    revertPirates();
  };
}
