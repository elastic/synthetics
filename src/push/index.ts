/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
import { readFile, writeFile } from 'fs/promises';
import { prompt } from 'enquirer';
import { bold, grey } from 'kleur/colors';
import {
  getLocalMonitors,
  buildMonitorSchema,
  diffMonitors as diffMonitorHashIDs,
  MonitorSchema,
} from './monitor';
import { ALLOWED_SCHEDULES, Monitor } from '../dsl/monitor';
import {
  progress,
  liveProgress,
  write,
  error,
  warn,
  indent,
  done,
  getMonitorManagementURL,
  THROTTLING_WARNING_MSG,
} from '../helpers';
import type { PushOptions, ProjectSettings } from '../common_types';
import { findSyntheticsConfig, readConfig } from '../config';
import {
  bulkDeleteMonitors,
  bulkGetMonitors,
  bulkPutMonitors,
  createMonitorsLegacy,
  CHUNK_SIZE,
} from './kibana_api';
import {
  getChunks,
  isBulkAPISupported,
  isLightweightMonitorSupported,
  logDiff,
} from './utils';

export async function push(monitors: Monitor[], options: PushOptions) {
  const duplicates = trackDuplicates(monitors);
  if (duplicates.size > 0) {
    throw error(formatDuplicateError(duplicates));
  }
  progress(
    `Pushing monitors for '${options.id}' project in kibana '${options.space}' space`
  );

  /**
   * Legacy API for kibana which does not support bulk operations
   */
  if (!isBulkAPISupported(options.kibanaVersion)) {
    return await pushLegacy(monitors, options);
  }

  const { monitors: remote } = await bulkGetMonitors(options);

  progress(`bundling ${monitors.length} monitors`);
  const schemas = await buildMonitorSchema(monitors, true);
  const local = getLocalMonitors(schemas);

  const { newIDs, changedIDs, removedIDs, unchangedIDs } = diffMonitorHashIDs(
    local,
    remote
  );
  logDiff(newIDs, changedIDs, removedIDs, unchangedIDs);

  const updatedMonitors = new Set<string>([...changedIDs, ...newIDs]);
  if (updatedMonitors.size > 0) {
    const chunks = getChunks(schemas, CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkPutMonitors(options, chunk),
        `creating or updating ${chunk.length} monitors`
      );
    }
  }

  if (removedIDs.size > 0) {
    if (updatedMonitors.size === 0 && unchangedIDs.size === 0) {
      await promptConfirmDeleteAll(options);
    }
    const chunks = getChunks(Array.from(removedIDs), CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkDeleteMonitors(options, chunk),
        `deleting ${chunk.length} monitors`
      );
    }
  }

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

async function promptConfirmDeleteAll(options: PushOptions) {
  write('');
  const { deleteAll } = await prompt<{ deleteAll: boolean }>({
    type: 'confirm',
    skip() {
      if (options.yes) {
        this.initial = process.env.TEST_OVERRIDE ?? true;
        return true;
      }
      return false;
    },
    name: 'deleteAll',
    message: `Pushing without any monitors will delete all monitors associated with the project.\n Do you want to continue?`,
    initial: false,
  });
  if (!deleteAll) {
    throw warn('Push command Aborted');
  }
}

function trackDuplicates(monitors: Monitor[]) {
  const monitorMap: Map<string, Monitor> = new Map();
  const duplicates: Set<Monitor> = new Set();
  for (const monitor of monitors) {
    const id = monitor.config.id;
    if (monitorMap.has(id)) {
      duplicates.add(monitorMap.get(id));
      duplicates.add(monitor);
    }
    monitorMap.set(id, monitor);
  }
  return duplicates;
}

export function formatDuplicateError(monitors: Set<Monitor>) {
  let outer = bold(`Aborted: Duplicate monitors found\n`);

  let inner = '';
  for (const monitor of monitors) {
    const { config, source } = monitor;
    inner += `* ${config.id} - ${source.file}:${source.line}:${source.column}\n`;
  }
  outer += indent(inner);
  return outer;
}

const INSTALLATION_HELP = `Run 'npx @elastic/synthetics init' to create project with default settings.`;

export async function loadSettings(configPath, ignoreMissing = false) {
  try {
    const config = await readConfig(
      process.env['NODE_ENV'] || 'development',
      configPath
    );
    // Missing config file, fake throw to capture as missing file
    if (Object.keys(config).length === 0) {
      throw '';
    }
    return config.project || ({} as ProjectSettings);
  } catch (e) {
    if (!ignoreMissing) {
      throw error(`Aborted (missing synthetics config file), Project not set up correctly.

${INSTALLATION_HELP}`);
    }
  }
}

export async function validatePush(
  opts: PushOptions,
  settings: ProjectSettings
) {
  validateSettings(opts);
  await catchIncorrectSettings(settings, opts);
}

export function validateSettings(opts: PushOptions) {
  const INVALID = 'Aborted. Invalid synthetics project settings.';

  let reason = '';
  if (!opts.id) {
    reason = `Set project id via
  - CLI '--id <id>'
  - Config file 'project.id' field`;
  } else if (!opts.locations && !opts.privateLocations) {
    reason = `Set default location for all monitors via
  - CLI '--locations <values...> or --privateLocations <values...>'
  - Config file 'monitors.locations' | 'monitors.privateLocations' field`;
  } else if (!opts.schedule) {
    reason = `Set default schedule in minutes for all monitors via
  - CLI '--schedule <mins>'
  - Config file 'monitors.schedule' field`;
  } else if (opts.schedule && !ALLOWED_SCHEDULES.includes(opts.schedule)) {
    reason = `Set default schedule(${
      opts.schedule
    }) to one of the allowed values - ${ALLOWED_SCHEDULES.join(',')}`;
  }

  if (!reason) return;

  throw error(`${INVALID}

${reason}

${INSTALLATION_HELP}`);
}

async function overrideSettings(
  configPath,
  oldValue: string,
  newValue: string
) {
  const cwd = process.cwd();
  configPath = configPath ?? (await findSyntheticsConfig(cwd, cwd));
  if (!configPath) {
    throw warn(`Unable to find synthetics config file: ${configPath}`);
  }
  const config = await readFile(configPath, 'utf-8');
  const updatedConfig = config.replace(
    `id: '${oldValue}'`,
    `id: '${newValue}'`
  );
  await writeFile(configPath, updatedConfig, 'utf-8');
}

export async function catchIncorrectSettings(
  settings: ProjectSettings,
  options: PushOptions
) {
  let override = !settings.id;
  if (settings.id && settings.id !== options.id) {
    // Add an extra line to make it easier to read the prompt
    write('');
    ({ override } = await prompt<{ override: boolean }>({
      type: 'confirm',
      name: 'override',
      skip() {
        if (options.yes) {
          this.initial = process.env.TEST_OVERRIDE ?? true;
          return true;
        }
        return false;
      },
      message: `Monitors were pushed under the '${settings.id}' project. Are you sure you want to push them under the new '${options.id}' (note that this will duplicate the monitors, the old ones being orphaned)`,
      initial: false,
    }));
    if (!override) {
      throw warn('Push command Aborted');
    }
  }
  if (override) {
    await overrideSettings(options.config, settings.id, options.id);
  }
}

export async function pushLegacy(monitors: Monitor[], options: PushOptions) {
  if (isLightweightMonitorSupported(monitors, options)) {
    throw error(
      `Aborted: Lightweight monitors are not supported in ${options.kibanaVersion}. Please upgrade to 8.5.0 or above.`
    );
  }

  let schemas: MonitorSchema[] = [];
  if (monitors.length > 0) {
    progress(`bundling ${monitors.length} monitors`);
    schemas = await buildMonitorSchema(monitors, false);
    const chunks = getChunks(schemas, 10);
    for (const chunk of chunks) {
      await liveProgress(
        createMonitorsLegacy({ schemas: chunk, keepStale: true, options }),
        `creating or updating ${chunk.length} monitors`
      );
    }
  } else {
    await promptConfirmDeleteAll(options);
  }
  await liveProgress(
    createMonitorsLegacy({ schemas, keepStale: false, options }),
    `deleting all stale monitors`
  );

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

// prints warning if any of the monitors has throttling settings enabled during push
export function warnIfThrottled(monitors: Monitor[]) {
  const throttled = monitors.some(monitor => monitor.config.throttling != null);
  if (throttled) {
    warn(THROTTLING_WARNING_MSG);
  }
}
