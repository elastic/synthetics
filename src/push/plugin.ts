/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { isAbsolute, dirname, extname, join } from 'path';
import fs from 'fs/promises';
import * as esbuild from 'esbuild';

// ROOT directory of the Package - /
const ROOT_DIR = join(__dirname, '..', '..');
// Source of the package - /src, /dist, etc.
const SOURCE_DIR = join(__dirname, '..');
// Node modules directory of the package - /node_modules
const SOURCE_NODE_MODULES = join(ROOT_DIR, 'node_modules');

export function commonOptions(): esbuild.BuildOptions {
  return {
    bundle: true,
    external: ['@elastic/synthetics'],
    minify: false,
    minifyIdentifiers: false,
    minifySyntax: false,
    minifyWhitespace: false,
    treeShaking: false,
    keepNames: false,
    platform: 'node',
    logLevel: 'silent',
    format: 'esm',
    write: false,
    outExtension: {
      '.js': '.js',
    },
  };
}

export type PluginData = {
  path: string;
  contents: string;
};
export type PluginCallback = (data: PluginData) => void;

export function MultiAssetPlugin(callback: PluginCallback): esbuild.Plugin {
  // Check that the path isn't in an external package by making sure it's at a standard
  // local filesystem location
  const isBare = (str: string) => {
    // Note that we use `isAbsolute` to handle UNC/windows style paths like C:\path\to\thing
    // This is not necessary for relative directories since `.\file` is not supported as an import
    // nor is `~/path/to/file`.
    if (isAbsolute(str) || str.startsWith('./') || str.startsWith('../')) {
      return true;
    }
    return false;
  };

  // If we're importing the @elastic/synthetics package
  // directly from source instead of using the fully
  // qualified name, we must skip it too. That's just
  // so it doesn't get bundled on tests or when we locally
  // refer to the package itself.
  const isLocalSynthetics = (entryPath: string) => {
    return entryPath.startsWith(SOURCE_DIR);
  };

  // When importing the local synthetics module directly
  // it may import its own local dependencies, so we must
  // make sure those will be resolved using Node's resolution
  // algorithm, as they're still "node_modules" that we must bundle
  const isLocalSyntheticsModule = (str: string) => {
    return str.startsWith(SOURCE_NODE_MODULES);
  };

  return {
    name: 'esbuild-multiasset-plugin',
    setup(build) {
      build.onResolve({ filter: /.*?/ }, async args => {
        // External and other packages need be marked external to
        // be removed from the bundle
        if (build.initialOptions.external?.includes(args.path)) {
          return {
            external: true,
          };
        }

        if (
          !isBare(args.path) ||
          args.importer.includes('/node_modules/') ||
          isLocalSyntheticsModule(args.importer)
        ) {
          return;
        }

        if (args.kind === 'entry-point') {
          return {
            path: args.path,
            namespace: 'journey',
          };
        }

        // If the modules are resolved locally, then
        // use the imported path to get full path
        const entryPath =
          join(dirname(args.importer), args.path) + extname(args.importer);

        if (isLocalSynthetics(entryPath)) {
          return { external: true };
        }

        // Spin off another build to copy over the imported modules without bundling
        const result = await esbuild.build({
          ...commonOptions(),
          entryPoints: {
            [entryPath]: entryPath,
          },
          bundle: false,
          external: [],
        });

        callback({
          path: entryPath,
          contents: result.outputFiles[0].text,
        });

        return {
          errors: result.errors,
          external: true,
        };
      });

      build.onLoad({ filter: /.*?/, namespace: 'journey' }, async args => {
        const contents = await fs.readFile(args.path, 'utf-8');
        callback({ path: args.path, contents });

        // Use correct loader for the journey entry path
        let loader: esbuild.Loader = 'default';
        const ext = extname(args.path).slice(1);
        if (ext === 'cjs' || ext === 'mjs') {
          loader = 'js';
        } else if (ext === 'cts' || ext === 'mts') {
          loader = 'ts';
        } else {
          loader = ext as esbuild.Loader;
        }
        return { contents, loader };
      });
    },
  };
}
