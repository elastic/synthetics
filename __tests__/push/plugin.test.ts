/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import * as esbuild from 'esbuild';
import { mkdir, rm, writeFile } from 'fs/promises';
import { join } from 'path';
import { commonOptions } from '../../src/core/transform';
import { SyntheticsBundlePlugin } from '../../src/push/plugin';

describe('SyntheticsBundlePlugin', () => {
  const PROJECT_DIR = join(__dirname, 'test-bundler');
  const journeyFile = join(PROJECT_DIR, 'bundle.journey.ts');

  beforeAll(async () => {
    await mkdir(PROJECT_DIR, { recursive: true });
  });

  afterAll(async () => {
    await rm(PROJECT_DIR, { recursive: true });
  });

  it('skip locally resolved synthetics package', async () => {
    // Should
    await writeFile(
      journeyFile,
      `import {journey, step, monitor} from '../../';
journey('journey 1', () => {
  monitor.use({ id: 'duplicate id' });
  step("step1", () => {})
});`
    );
    const result = await esbuild.build({
      ...commonOptions(),
      bundle: false,
      sourcemap: false,
      write: false,
      entryPoints: [journeyFile],
      plugins: [SyntheticsBundlePlugin()],
    });
    expect(result.outputFiles[0].text).toMatchSnapshot();
  });
});
