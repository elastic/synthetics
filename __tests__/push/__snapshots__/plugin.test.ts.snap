// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyntheticsBundlePlugin skip locally resolved synthetics package 1`] = `
"var import__ = require("../../");
(0, import__.journey)("journey 1", () => {
  import__.monitor.use({ id: "duplicate id" });
  (0, import__.step)("step1", () => {
  });
});
"
`;
